### PR TITLE
Add an intrinsic implementation of `type(object)` function

### PR DESCRIPTION
**Pull request by kmr-srbh**
Currently the logic is simple - get the type of the passed argument using `ASRUtils::type_to_str_python(*ttype_t)` and return it as a StringConstant. I request guidance to improve this PR.

### Get type of the object
- Variables
```python
integer_list: list[i32] = [1, 2, 3, 4]
print(type(integer_list))
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
<type 'list[i32]'>
```
- Constants
```python
print(type("Hello, LPython"))
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
<type 'str'>
```
### Compare object types (requires #2591 to be merged)
```python
integer: i32 = 1
integer_list: list[i32] = [1, 2, 3, 4, 5]
print(type(integer) == type(integer_list))
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
False
```

### CODE REVIEW

**Comment by kmr-srbh** [comment]
The reason which causes the above 2 tests to fail is also the one I need help with. The argument `Vec<ASR::expr_t*> args` remains unused as I get the type directly through `ASR::ttype_t* t1`. @Thirumalai-Shaktivel please have a look into the PR.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 1]
```diff
@@ -1160,6 +1162,53 @@ static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
...
+namespace ObjectType {
+
+     static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x, diag::Diagnostics& diagnostics) {
+        ASRUtils::require_impl(x.n_args != 1,
+            "ASR Verify: type() takes only 1 argument `object`",
+            x.base.base.loc, diagnostics);
+    }
+
+    static ASR::expr_t *eval_ObjectType(Allocator &al, const Location &loc,
+            ASR::ttype_t* t1, Vec<ASR::expr_t*>& args) {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:1175]: Here is the issue:
```suggestion
            ASR::ttype_t* t1, Vec<ASR::expr_t*>& /*args*/) {
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 2]
```diff
@@ -1160,6 +1162,53 @@ static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
...
+    static inline ASR::asr_t* create_ObjectType(Allocator& al, const Location& loc,
+            Vec<ASR::expr_t*>& args,
+            const std::function<void (const std::string &, const Location &)> err) {
+        if (args.size() > 1) {
+            err("type() takes only 1 argument `object`", loc);
+        }
+
+        Vec<ASR::expr_t *> arg_values;
+        arg_values.reserve(al, 1);
+        arg_values.push_back(al, expr_value(args[0]));
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:1189]:
```suggestion
        Vec<ASR::expr_t *> arg_values;
```
We don't use the expr_value(args[0]) here, as the output is dependent on the argument type.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 3]
```diff
@@ -1160,6 +1162,53 @@ static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
...
+        std::string object_type = "<type '" + ASRUtils::type_to_str_python(t1) + "'>";
+        return StringConstant(object_type, character(object_type.length()));
+    }
+
+    static inline ASR::asr_t* create_ObjectType(Allocator& al, const Location& loc,
+            Vec<ASR::expr_t*>& args,
+            const std::function<void (const std::string &, const Location &)> err) {
+        if (args.size() > 1) {
+            err("type() takes only 1 argument `object`", loc);
+        }
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:1185]:
```suggestion
        if (args.size() == 1) {
            err("type() takes exactly 1 argument `object` for now", loc);
        }
```

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 4]
```diff
@@ -7487,6 +7487,9 @@ we will have to use something else.
             throw SemanticError("Only Name or Attribute type supported in Call",
                 x.base.base.loc);
         }
+        if (call_name == "type" && x.n_args == 0) {
+            throw SemanticError("type() takes 1 required argument `object`", x.base.base.loc);
+        }
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/semantics/python_ast_to_asr.cpp:7492]:
```suggestion
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 5]
```diff
@@ -1160,6 +1162,53 @@ static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
...
+            Vec<ASR::call_arg_t>& new_args, int64_t /*overload_id*/) {
+        declare_basic_variables("_lcompilers_objectType_" + type_to_str_python(arg_types[0]));
+        fill_func_arg("a", arg_types[0]);
+        auto result = declare(fn_name, return_type, ReturnVar);
+
+        ASR::symbol_t *f_sym = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
+        scope->add_symbol(fn_name, f_sym);
+        return b.Call(f_sym, new_args, return_type, nullptr);
+    }
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:1207]: I think the compile time value will always be used, so there is no need for instantiate function.
```suggestion
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 6]
```diff
@@ -3625,6 +3674,8 @@ namespace IntrinsicScalarFunctionRegistry {
     static const std::map<int64_t,
         std::tuple<impl_function,
                    verify_function>>& intrinsic_function_by_id_db = {
+        {static_cast<int64_t>(IntrinsicScalarFunctions::ObjectType),
+            {&ObjectType::instantiate_ObjectType, &UnaryIntrinsicFunction::verify_args}},
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:3678]:
```suggestion
            {nullptr, &UnaryIntrinsicFunction::verify_args}},
```

File: src/libasr/pass/intrinsic_function_registry.h (hunk 1 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h]: Thanks a lot for this! :)

File: src/libasr/pass/intrinsic_function_registry.h (hunk 3 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:1185]: CPython has the `type()` overloaded. One of them takes just 1 argument `object`. While discussing with you earlier in the day, I was mistakenly referring to the other function which creates a new type - `type(name, dict, **kwargs)`.

This should remain unchanged I think.

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:1185]: According to the implement in this PR, we use or handle or perform operation on the 1st argument, so I suggested the above change. Later, we can handle more arguments.

**Comment by Thirumalai-Shaktivel** [comment]
Also, Let's add some tests.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 4 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]: This check is required as an unhandled exception is thrown for an empty function call.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 3 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:1185]: @Thirumalai-Shaktivel did you mean to say `args.size() != 1`?

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 4 above)
Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]: I think every semantics checks related to `type(..)` needs to be handled inside the `create_ObjectType` function.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 3 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:1185]: Yup!

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 4 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]: It's not @Thirumalai-Shaktivel. That's why I had to add this here in the call visitor :(

Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]: Share an minimum reproducible example, I will check it out.

Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]: Do you mean for the following example?
```py
print(type())
```

CPython and LPython both throws following errors:
```console
$ python examples/expr2.py 
Traceback (most recent call last):
  File "/Users/thirumalai/Open_Source/lpython/examples/expr2.py", line 1, in <module>
    print(type())
TypeError: type() takes 1 or 3 arguments
$ lpython examples/expr2.py 
semantic error: type() takes 1 required argument `object`
 --> examples/expr2.py:1:7
  |
1 | print(type())
  |       ^^^^^^ 

```

Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]:
```python
print(type())
```

Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]: My bad, yea, it throws an exception:
```console
....
  File "/Users/thirumalai/Open_Source/lpython/src/libasr/asr.h", line 49205
    LCOMPILERS_ASSERT(f != nullptr);
AssertFailed: f != nullptr
```

Comment (Thirumalai-Shaktivel) [reply src/lpython/semantics/python_ast_to_asr.cpp:7492]: Apply the following diff to fix the bug:
```diff
diff --git a/src/lpython/semantics/python_ast_to_asr.cpp b/src/lpython/
semantics/python_ast_to_asr.cpp
index 714349d7c..9ae12641d 100644
--- a/src/lpython/semantics/python_ast_to_asr.cpp
+++ b/src/lpython/semantics/python_ast_to_asr.cpp
@@ -7487,9 +7487,6 @@ we will have to use something else.
             throw SemanticError("Only Name or Attribute type supported
 in Call",
                 x.base.base.loc);
         }
-        if (call_name == "type" && x.n_args == 0) {
-            throw SemanticError("type() takes 1 required argument `obj
ect`", x.base.base.loc);
-        }
 
         ASR::symbol_t *s = current_scope->resolve_symbol(call_name);
         if( s && ASR::is_a<ASR::Module_t>(*s) ) {
@@ -7522,7 +7519,7 @@ we will have to use something else.
                 }
                 Vec<ASR::expr_t*> args_; args_.reserve(al, x.n_args);
                 visit_expr_list(x.m_args, x.n_args, args_);
-                if (ASRUtils::is_array(ASRUtils::expr_type(args_[0])) 
&&
+                if (x.n_args && ASRUtils::is_array(ASRUtils::expr_type
(args_[0])) &&
                     imported_functions[call_name] == "math" ) {
                     throw SemanticError("Function '" + call_name + "' 
does not accept vector values",
                         x.base.base.loc);
```

**Comment by kmr-srbh** [comment]
> Also, Let's add some tests.

There is a small issue. The return type of `type()` is currently the same as the passed argument. How do we go about fixing it?

**Comment by Thirumalai-Shaktivel** [comment]
Minimum reproducible example (MRE), please?

**Comment by kmr-srbh** [comment]
> Minimum reproducible example (MRE), please?

```python
l: list[i32] = [1, 2, 3, 4, 5]
t: str = type(l)
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 7]
```diff
@@ -1160,6 +1162,39 @@ static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
...
+            const std::function<void (const std::string &, const Location &)> err) {
+        if (args.size() != 1) {
+            err("type() takes exactly 1 argument `object` for now", loc);
+        }
+
+        Vec<ASR::expr_t *> arg_values;
+        ASR::expr_t *m_value = eval_ObjectType(al, loc, expr_type(args[0]), arg_values);
+        return ASR::make_IntrinsicScalarFunction_t(al, loc,
+            static_cast<int64_t>(IntrinsicScalarFunctions::ObjectType),
+            args.p, args.n, 0, ASRUtils::expr_type(args[0]), m_value);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:1191]:
```suggestion
            args.p, args.n, 0, ASRUtils::expr_type(m_value), m_value);
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 8]
```diff
@@ -1160,6 +1162,39 @@ static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
 
 } // namespace BinaryIntrinsicFunction
 
+
+namespace ObjectType {
+
+     static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x, diag::Diagnostics& diagnostics) {
+        ASRUtils::require_impl(x.n_args != 1,
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:1169]:
```suggestion
        ASRUtils::require_impl(x.n_args == 1,
```

File: src/libasr/pass/intrinsic_function_registry.h [hunk 9]
```diff
@@ -3625,6 +3660,8 @@ namespace IntrinsicScalarFunctionRegistry {
     static const std::map<int64_t,
         std::tuple<impl_function,
                    verify_function>>& intrinsic_function_by_id_db = {
+        {static_cast<int64_t>(IntrinsicScalarFunctions::ObjectType),
+            {nullptr, &UnaryIntrinsicFunction::verify_args}},
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:3664]:
```suggestion
            {nullptr, &ObjectType::verify_args}},
```

**Review by Thirumalai-Shaktivel — COMMENTED** [review]
I think the following code output would differ with CPython, we have to fix it before merging this PR.
Rest looks good.
```py
print(type(type(1)))
```

File: src/libasr/pass/intrinsic_function_registry.h (hunk 7 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:1191]: This fails withe the ASR Verify error stated above.

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:1191]: Apply my other suggestions, the ASR Verify error would be fixed.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 8 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:1169]: Hmm.. this check was really confusing. It was a bit difficult to understand what we were checking. Thanks a lot for this! :+1:

File: src/libasr/pass/intrinsic_function_registry.h (hunk 7 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:1191]: I am in the process 😃.

Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h]: @Thirumalai-Shaktivel I am interested to know the reason this worked. I guess it was because of this line - `{nullptr, &ObjectType::verify_args}}`?

**Comment by kmr-srbh** [comment]
> I think the following code output would differ with CPython, we have to fix it before merging this PR. Rest looks good.
> 
> ```python
> print(type(type(1)))
> ```

Going the `print()` way?

File: src/libasr/pass/intrinsic_function_registry.h (hunk 7 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h]: Yes! It's because of the function `UnaryIntrinsicFunction::verify_args` which is used for the intrinsics that have same type for the input and output.

**Comment by Thirumalai-Shaktivel** [comment]
Nope, I meant the case: type(type(1))
Here CPython prints as: `<class 'type'>`, But LPython using this PR changes, prints as `<type 'str'>`.
We need to fix this.

**Comment by kmr-srbh** [comment]
> Nope, I meant the case: type(type(1)) Here CPython prints as: `<class 'type'>`, But LPython using this PR changes, prints as `<type 'str'>`. We need to fix this.

But this will always be the case until we create a specific type for `type()`. Imagine `type(type(type(1)))`.

**Comment by Thirumalai-Shaktivel** [comment]
Check if the args[0] is IntrinsicScalarFunciton and down_cast it and check if the m_intrinsic_id is static_cast<IntrinsicScalarFunctions::ObjectType>, if so then store the output as <class 'type'>. Otherwise do eval_ObjectType, it would work for any number of nested type(..).

Later in the future, if we find any other different output or cases, we will modify it accordingly.

**Comment by kmr-srbh** [comment]
> Check if the `args[0]` is `IntrinsicScalarFunciton` and `down_cast` it and check if the `m_intrinsic_id` is `static_cast<IntrinsicScalarFunctions::ObjectType>`, if so then store the output as `<class 'type'>`. Otherwise do `eval_ObjectType`, it would work for any number of nested `type(..)`.
> 
> Later in the future, if we find any other different output or cases, we will modify it accordingly.

Hmm, this seems a good idea. Let's test it out. :+1:

**Comment by kmr-srbh** [comment]
I am getting this error now. I think it is because of the clean build we did for the `isspace()` branch. 
```console
CMake Error at CMakeLists.txt:10 (file):
  file STRINGS file "/home/saurabh-kumar/Projects/System/lpython/version"
  cannot be read.
.
.
.
-- Configuring incomplete, errors occurred!
```

**Comment by Thirumalai-Shaktivel** [comment]
What command did you use to build `lpython`?

**Comment by kmr-srbh** [comment]
> Which command did you use to build `lpython`?

When this error occurred now, I followed the complete process which we do on a fresh build. As it turns out, only `./build1.sh` threw the error. Other commands did not do anything.

**Comment by Thirumalai-Shaktivel** [comment]
Before `./build1.sh`, we have to run `./build0.sh` to generate all the required files.

Or you can just run `./build.sh`

**Comment by kmr-srbh** [comment]
Thanks @Thirumalai-Shaktivel it works now. It was an issue with the tag which had previously caused `./build0.sh` to fail. What is the correct way of using tags here at LPython? Always using the default one?

**Comment by Thirumalai-Shaktivel** [comment]
Yes, I think git automatically picks the correct tag, see ci/version.sh

**Comment by kmr-srbh** [comment]
I am pushing the tests.

**Comment by kmr-srbh** [comment]
Thanks a lot for your support and guidance @Thirumalai-Shaktivel! Without you I could not have made it this far. Thank you very much!

**Comment by kmr-srbh** [comment]
@Thirumalai-Shaktivel do you think this PR is ready?

File: src/libasr/pass/intrinsic_function_registry.h [hunk 10]
```diff
@@ -1160,6 +1162,49 @@ static inline void verify_args(const ASR::IntrinsicScalarFunction_t& x,
...
+        if (args.size() != 1) {
+            err("type() takes exactly 1 argument `object` for now", loc);
+        }
+        ASR::expr_t *m_value = nullptr;
+        Vec<ASR::expr_t *> arg_values;
+       
+        if (ASR::is_a<ASR::IntrinsicScalarFunction_t>(*args[0])) {
+            ASR::IntrinsicScalarFunction_t *object = ASR::down_cast<ASR::IntrinsicScalarFunction_t>(args[0]);
+            if (static_cast<IntrinsicScalarFunctions>(object->m_intrinsic_id) == IntrinsicScalarFunctions::ObjectType) {
+                m_value = StringConstant("<type 'type'>", character(13)); // 13 is the length of the string "<type 'type'>"
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:1192]:
```suggestion
                m_value = StringConstant("<class 'type'>", character(13)); // 13 is the length of the string "<type 'type'>"
```
```py
>>> type(type(1))
<class 'type'>
```

File: tests/tests.toml [hunk 11]
```diff
@@ -410,6 +410,10 @@ asr = true
 filename = "../integration_tests/test_builtin_oct.py"
 asr = true
 
+[[test]]
+filename = "../integration_tests/test_builtin_type.py"
+asr = true
+
```
Comment (Thirumalai-Shaktivel) [inline tests/tests.toml:416]:
```suggestion
```

Tests this using `integration_tests/CMakeLists`

test: llvm, cpython

Comment (kmr-srbh) [reply tests/tests.toml:416]: Okay.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 10 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:1192]: I had intentionally not used the word 'class' as we do not currently support it fully. Should we write it here?

File: tests/tests.toml (hunk 11 above)
Comment (kmr-srbh) [reply tests/tests.toml]: The cpython test fails because the `type()` has a different return type there.

File: src/libasr/pass/intrinsic_function_registry.h (hunk 10 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:1192]: Nope, we have to make sure both CPython and LPython generates the same output.
So, I suggested the above changes and test this both in CPython and LPython(llvm).

Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_function_registry.h:1192]: I am really sorry for the confusion. The return type of `type()` in CPython is of type - `type`, whereas here we are using a string for it. They would not be similar.

I request guidance on how do we go about this @Thirumalai-Shaktivel.

**Comment by Thirumalai-Shaktivel** [comment]
Apply the following diff:
Let's test only the data types for now; later, we can handle it based on the requirements.
```diff
diff --git a/integration_tests/test_builtin_type.py b/integration_tests/test_builtin_type.py
index dea42bc74..961ed245c 100644
--- a/integration_tests/test_builtin_type.py
+++ b/integration_tests/test_builtin_type.py
@@ -1,19 +1,27 @@
-from lpython import i32, f64, list, str, dict, Const
+from lpython import i32, f64, Const

 def test_builtin_type():
     i: i32 = 42
+    f: f64 = 64.
     s: str = "Hello, LPython!"
     l: list[i32] = [1, 2, 3, 4, 5]
     d: dict[str, i32] = {"a": 1, "b": 2, "c": 3}
-    CONST_LIST: Const[list[f64]] = [12.22, 14.63, 33.82, 19.18]
+    res: str = ""

-    assert type(i) == "<type 'i32'>"
-    assert type(s) == "<type 'str'>"
-    assert type(l) == "<type 'list[i32]'>"
-    assert type(d) == "<type 'dict[str, i32]'>"
-    assert type(CONST_LIST) == "<type 'Const[list[f64]]'>"
-
-    assert type(type(i)) == "<type 'type'>"
-    assert type(type(CONST_LIST)) == "<type 'type'>"
+    res = str(type(i))
+    print(res)
+    assert res == "<class 'int'>"
+    res = str(type(f))
+    print(res)
+    assert res == "<class 'float'>"
+    res = str(type(s))
+    print(res)
+    assert res == "<class 'str'>"
+    res = str(type(l))
+    print(res)
+    assert res == "<class 'list'>"
+    res = str(type(d))
+    print(res)
+    assert res == "<class 'dict'>"

 test_builtin_type()
diff --git a/src/libasr/pass/intrinsic_function_registry.h b/src/libasr/pass/intrinsic_function_registry.h
index e70524b62..19d219f53 100644
--- a/src/libasr/pass/intrinsic_function_registry.h
+++ b/src/libasr/pass/intrinsic_function_registry.h
@@ -1173,7 +1173,24 @@ namespace ObjectType {

     static ASR::expr_t *eval_ObjectType(Allocator &al, const Location &loc,
             ASR::ttype_t* t1, Vec<ASR::expr_t*>& /*args*/) {
-        std::string object_type = "<type '" + ASRUtils::type_to_str_python(t1) + "'>";
+        std::string object_type = "<class '";
+        switch (t1->type) {
+            case ASR::ttypeType::Integer : {
+                object_type += "int"; break;
+            } case ASR::ttypeType::Real : {
+                object_type += "float"; break;
+            } case ASR::ttypeType::Character : {
+                object_type += "str"; break;
+            } case ASR::ttypeType::List : {
+                object_type += "list"; break;
+            } case ASR::ttypeType::Dict : {
+                object_type += "dict"; break;
+            } default: {
+                LCOMPILERS_ASSERT_MSG(false, "Unsupported type");
+                break;
+            }
+        }
+        object_type += "'>";
         return StringConstant(object_type, character(object_type.length()));
     }

@@ -1185,17 +1202,8 @@ namespace ObjectType {
         }
         ASR::expr_t *m_value = nullptr;
         Vec<ASR::expr_t *> arg_values;
-
-        if (ASR::is_a<ASR::IntrinsicScalarFunction_t>(*args[0])) {
-            ASR::IntrinsicScalarFunction_t *object = ASR::down_cast<ASR::IntrinsicScalarFunction_t>(args[0]);
-            if (static_cast<IntrinsicScalarFunctions>(object->m_intrinsic_id) == IntrinsicScalarFunctions::ObjectType) {
-                m_value = StringConstant("<type 'type'>", character(13)); // 13 is the length of the string "<type 'type'>"
-            }
-        }
-        else {
-            m_value = eval_ObjectType(al, loc, expr_type(args[0]), arg_values);
-        }
-
+        m_value = eval_ObjectType(al, loc, expr_type(args[0]), arg_values);
+
         return ASR::make_IntrinsicScalarFunction_t(al, loc,
             static_cast<int64_t>(IntrinsicScalarFunctions::ObjectType),
             args.p, args.n, 0, ASRUtils::expr_type(m_value), m_value);
```

File: tests/reference/asr-test_builtin_type-f795c34.json [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+    "infile": "tests/../integration_tests/test_builtin_type.py",
+    "infile_hash": "81fc4815e7adc95e05982852652f2cb26658ae4e46d7b635d0d5f580",
+    "outfile": null,
+    "outfile_hash": null,
+    "stdout": "asr-test_builtin_type-f795c34.stdout",
+    "stdout_hash": "6e813b30a83bfc41b0ca67e40451b1b81edd29b56029334eb7bba89e",
+    "stderr": null,
+    "stderr_hash": null,
+    "returncode": 0
+}
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/asr-test_builtin_type-f795c34.json:13]: Remove these two files from tests/reference

**Comment by Thirumalai-Shaktivel** [comment]
@Shaikh-Ubaid, do you have any design suggestions here?

File: integration_tests/test_builtin_type.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+from lpython import i32, f64, Const
```
Comment (Shaikh-Ubaid) [inline integration_tests/test_builtin_type.py:1]: If `Const` is unused, remove it.

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 14]
```diff
@@ -3040,6 +3040,7 @@ PyMODINIT_FUNC PyInit_lpython_module_)" + fn_name + R"((void) {
         std::string out;
         std::string indent(4, ' ');
         switch (x.m_intrinsic_id) {
+            SET_INTRINSIC_NAME(ObjectType, "type");
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/asr_to_c_cpp.h:3043]: Does the integration test added in this PR work with the `c` backend. If it works, then enable the `c` backend as well. If not, then I think we need not update the `c` backend here.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 15]
```diff
@@ -7586,7 +7586,8 @@ we will have to use something else.
                 }
                 Vec<ASR::expr_t*> args_; args_.reserve(al, x.n_args);
                 visit_expr_list(x.m_args, x.n_args, args_);
-                if (ASRUtils::is_array(ASRUtils::expr_type(args_[0])) &&
+                
+                if (x.n_args && ASRUtils::is_array(ASRUtils::expr_type(args_[0])) &&
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:7590]: Is this change needed for this PR? What specific case does it fix?

If this change is not directly related to this PR, but was related to some issue found while working on this PR, then I suggest reporting the specific issue found and then sending a separate PR fixing it.

**Review by Shaikh-Ubaid — COMMENTED** [review]
I shared some comments. Overall the implementation looks in the right direction.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 15 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp:7590]: It avoids an exception when the function is called with no parameters.

File: src/libasr/codegen/asr_to_c_cpp.h (hunk 14 above)
Comment (kmr-srbh) [reply src/libasr/codegen/asr_to_c_cpp.h:3043]: It works. I have added the `c` back-end. :+1:

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 15 above)
Comment (Shaikh-Ubaid) [reply src/lpython/semantics/python_ast_to_asr.cpp:7590]: >It avoids an exception when the function is called with no parameters.

Do you mean when the `type()` function is called with no paramaters?

File: src/libasr/codegen/asr_to_c_cpp.h (hunk 14 above)
Comment (Shaikh-Ubaid) [reply src/libasr/codegen/asr_to_c_cpp.h:3043]: Perfect.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 15 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp:7590]: Yes. It is due to this check that we are able to check for `type()` called with no parameters in the `create_ObjectType()`. Without that check we have an unhandled exception.

Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:7590]: We can make it more explicit.
```suggestion
                if (x.n_args > 0 && ASRUtils::is_array(ASRUtils::expr_type(args_[0])) &&
```

**Review by Shaikh-Ubaid — COMMENTED** [review]
We just need to fix https://github.com/lcompilers/lpython/pull/2592#discussion_r1527573997 and https://github.com/lcompilers/lpython/pull/2592#discussion_r1527589247. I think then this would be good to merge.

File: integration_tests/test_builtin_type.py (hunk 13 above)
Comment (kmr-srbh) [reply integration_tests/test_builtin_type.py]: It was used before, but when we decided to make it work exactly like Python, we removed checks for `Const` types.

**Review by Shaikh-Ubaid — APPROVED** [review]
It looks good! Thanks!
